#25: Introduced break-after attribute on sequences; added values 'page' and 'sheet' to existing break-before attributes

PaulRambags opened the pull request:
Closes #25.

@bertfrees @kalaspuffar please review this pull request

You may also have a look at this temporary available [preview](https://rambags.home.xs4all.nl/obfl/obfl-specification.html).

PaulRambags commented on src/obfl-specification.html:
```diff
@@ -343,13 +343,56 @@ <h3 id="L5025">Auxiliary Flows</h3>
+  <dl>
+    <dt>'volume'</dt>
+      <dd>continue in a new volume with or after this sequence,
+        respectively.</dd>
```
I am not sure about this myself, because 'volume' is not an allowed value in any `break-after`. I have put it like this just to have a common, global overview of all the values that are available in `break-before` or `break-after`. @bertfrees what do you think?

bertfrees replied:
Since this is a reference document, which should be aimed at finding information efficiently and not so much at providing a pleasant reading experience, I would opt for simply explaining the allowed values for each element separately, thereby repeating stuff where needed. Especially in this case where the allowed values, and even their meanings, depend on the context (element/attribute). I think this list, at least the way it is currently worded, does more bad than good.

I see what you're trying to do here though. In the paragraph below you explain how the attributes interact, and for this it is indeed required to already have a basic understanding of which values exist. Not sure how to solve this best.

bertfrees replied:
Also, nothing is stopping us from adding `volume` to the `break-after` values. Would be good for consistency.

PaulRambags replied:
I've added `volume` to `break-after`.

bertfrees commented:
@PaulRambags Do you happen to have a preview link again? :-)

PaulRambags commented:
> @PaulRambags Do you happen to have a preview link again? :-)

@bertfrees Yes. Please see above, at the first comment in this conversation. ;-)

bertfrees commented:
🤦

bertfrees commented on src/obfl-specification.html:
```diff
@@ -343,13 +343,56 @@ <h3 id="L5025">Auxiliary Flows</h3>
+  paginated. The default behavior is that there is a back side and it is counted
+  but not <em>paginated</em>.</p>
+
+<ins datetime="20200612T15:00:00">The last page of a sequence where duplex is
+enabled, may not be paginated and may not be counted.</ins>
+<p></p>
```
It is only now that I noticed the "Backwards Incompatible Changes" section. Are all of these changes actually backwards incompatible? For example, the "external-reference" addition. Also the changes in this PR are not backwards incompatible as far as I know. At least in my proposal, I deliberately kept the default behavior the same.

PaulRambags replied:
You're right. It is backwards compatible. You don't have to use the `break-after` element and the `page` value, which makes it perfectly backward compatible. I will remove this `<ins ... />`.

bertfrees commented on src/obfl-specification.html:
```diff
@@ -343,13 +343,56 @@ <h3 id="L5025">Auxiliary Flows</h3>
+      <dd>continue in a new volume with or after this sequence,
+        respectively.</dd>
+    <dt>'sheet'</dt>
+      <dd>continue on a new right-hand page with or after this sequence,
+        respectively. If this results in a page without content, don't skip the
```
> continue on a new right-hand page with or after this sequence, respectively.

I found it hard to understand this sentence.
- "with or after": strange wording
- "respectively": refers to the "break-before and break-after" in the first paragraph, but this is not immediately obvious.

Maybe just simplify it to "inserts a sheet break"? Because in the first paragraph you already explain _where_ breaks are inserted, so here it should be enough to say _which kind_ of breaks. In the "break-before" and "break-after" you can explain in somewhat more detail.

PaulRambags replied:
I have simplified the description of the values.

bertfrees commented on src/obfl-specification.html:
```diff
@@ -343,13 +343,56 @@ <h3 id="L5025">Auxiliary Flows</h3>
+        header or footer. This behavior is consistent with break-before="sheet"
+        on a block.</dd>
+    <dt>'page'</dt>
+      <dd>continue on a new page with or after this sequence, respectively,
+        which may be a left-hand page. If there is a transition from duplex to
+        simplex or visa versa, or the page dimensions don't match, the current
+        or new sequence needs to start on a right-hand page.</dd>
```
Simplify to "insert a page break"?

PaulRambags replied:
Done. With some extra explanation that this value can be used to produce media without blank pages.

bertfrees commented on src/obfl-specification.html:
```diff
@@ -1027,7 +1070,11 @@ <h4 id="L906" class="include">sequence</h4>
+          <dd>One of 'auto' (default), 'sheet', 'page' or 'volume'.</dd>
+		<dt>break-after [optional]</dt>
+          <dd class="no-markup">Inserts a break after the sequence has ended.</dd>
+          <dd>This property only applies to text body sequences.</dd> 
```
> This property only applies to text body sequences.

Why this limitation? Actually I don't think we can do without this attribute on sequences in pre-content.

PaulRambags replied:
Removed.

PaulRambags replied:
@bertfrees Although I have removed the "This property only applies to text body sequences." I think it is kind of strange that this limitation is present in the `break-before` attribute. I don't see why the limitation should be present on `break-before` but not on `break-after`.

I don't want to mention that this attribute is not needed in sequences in pre-content. It is also not mentioned that `break-before` is not needed on sequences in post-content.

bertfrees replied:
Ah, you mean that sentence was already there for `break-before`? OK I understand. It was  referring to the value "volume" (which used to be the only value), which indeed does not make sense to be used inside `pre-content` and `post-content`. So I guess we should keep the sentence but move it to the description of the "volume" value. I also wonder whether it would be worth having this reflected in the schema?

bertfrees replied:
@PaulRambags What do you think about the schema? Should we make sure you can not use `volume` inside `pre-content`?

PaulRambags replied:
I have been thinking about this one. It is certainly possible to do so but it makes the schema more complicated. My personal opinion is that a schema should be as strict as possible, but it should still be well-structured. I will do a try.

PaulRambags replied:
Done.

bertfrees commented on src/obfl-specification.html:
```diff
@@ -1071,6 +1118,14 @@ <h4 class="include" id="L11470">dynamic-sequence</h4>
+          <dd>One of 'sheet' or 'page'.</dd>
+		<dt>break-after [optional]</dt>
+          <dd class="no-markup">Inserts a break after the sequence has ended.</dd>
+          <dd>This property only applies to text body sequences.</dd> 
```
See comment above. Also this doesn't even make sense, dynamic-sequence can only appear within pre- or post-content.

PaulRambags replied:
@bertfrees Why is there a `break-before` attibute on dynamic sequences?

I have added the `break-after` here because there is also a `break-before`.

Why does it not make sense? Suppose you have a post-content with `break-after="page"` and the post-content ends on a right page. Then the next volume starts at a left page, instead of the default behavior.

bertfrees replied:
> Why is there a break-before attibute on dynamic sequences?

You mean why we are adding it? Because we want to control the type of page break. Not only for `sequence` but also for `dynamic-sequence` (and `toc-sequence`). So yes, both `break-before` and `break-after` make sense.

> Why does it not make sense?

I meant that the sentence "This property only applies to text body sequences" does not make sense in the context of `dynamic-sequence`, because a `dynamic-sequence` can never appear in the body.

bertfrees commented on src/obfl-specification.html:
```diff
@@ -343,13 +343,56 @@ <h3 id="L5025">Auxiliary Flows</h3>
-the back side is counted, but not <em>paginated</em>.</p>
+<p class="markup">In a sequence where duplex is enabled, both sides of sheets
+  are counted and paginated. However, if the last sheet of a duplex sequence
+  only has content on the front side, it depends on the values of the optional
+  attributes break-before and break-after on sequence, dynamic-sequence and
+  toc-sequence whether there is a back side, and if so, whether it is counted or
+  paginated. The default behavior is that there is a back side and it is counted
+  but not <em>paginated</em>.</p>
```
- "optional attributes": that the attributes are optional is not so relevant
- "whether there is a back side": I think I understand what you mean with this. No back side is when the next sequence starts on the back side, right? For someone new it's hard to infer that from this explanation I think.
- "whether it is counted": are there situations where it is _not_ counted?

PaulRambags replied:
I have modified this section according to your suggestions.

PaulRambags commented:
> I've read the newly added documentation multiple times. It might be that the topic is hard to cover, but I think it's hard to parse what the different behaviors would end up as. The additions are technically correct but could be improved by adding some examples to explain different expected scenarios.

I have modified the description to make it clearer, simpler and better understandable. I also added an explanation at the 'page' value that this value can be used to produce media without blank pages. I hope it is better now. @kalaspuffar Please let me know if it is still unclear or can be improved in one way or another.

PaulRambags commented:
@bertfrees @kalaspuffar I have made changes according to your valuable remarks! Please review again.

You may also have a look at this temporary available [preview](https://rambags.home.xs4all.nl/obfl/obfl-specification.html#L5024).

bertfrees commented on src/obfl-specification.html:
```diff
@@ -1071,6 +1112,12 @@ <h4 class="include" id="L11470">dynamic-sequence</h4>
             information.</dd>
+		<dt>break-before [optional]</dt>
+          <dd class="no-markup">Inserts a break before the sequence begins.</dd>
+          <dd>One of 'sheet' or 'page'.</dd>
```
Did you forget 'auto' here?

PaulRambags replied:
I did! Thanks.

I have added 'auto' here and also at another place where it was missing at the `break-before`.

bertfrees commented:
@PaulRambags Thanks, much more understandable now.

When I look at the preview I notice that something has gone wrong with the automatic creation of cross-reference. "before" and "after" are converted to links to the `before` and `after` elements which have nothing to do with it.

Another thing is that when I am reading the "toc-sequence" section I get no explanation of the "break-before" values. I'd have to have previously read the "Duplex" section, but I don't think you can just assume that. You have chosen to collect these explanations in a single section so as to not having to repeat it at multiple places, which is fine, but this way it's a bit hidden away. Some hint that you can find the information there would be good.

bertfrees commented on src/obfl-specification.html:
```diff
@@ -343,13 +343,52 @@ <h3 id="L5025">Auxiliary Flows</h3>
+  </dl>
+</div>
+
+<p class="markup">The break-before and break-after attributes cover the cases
+  where you transition from the last sequence of the pre-content to the body
+  when the body is a resumption of a sequence in the previous volume. In cases
+  where a sequence is interrupted at the end of a volume and is followed by the
+  first sequence in the post-content, it is the break-before value of that new
+  sequence that matters. In cases where a sequence ends and is followed by a new
```
This is quite literally taken from my GH issue. But my text was more to argue why we need two attributes. It's not so important for the reader to understand the reasoning behind it. I would shorten it as follows:

> When a sequence ends and is followed by a new sequence, both the break-after value of the sequence that ended and the break-before value of the new sequence must be taken into account: 'volume' dominates over all the other values, 'sheet' dominates over 'page' and 'auto', and 'page' dominates over 'auto'. Note that a sequence does not 'end' when it is interrupted by a series of post-content and/or pre-content sequences (i.e. at a volume break), and likewise it does not 'start' when it is resumed in the next volume.

I think this is slightly more to the point.

PaulRambags replied:
Done.

PaulRambags commented:
@bertfrees 

> When I look at the preview I notice that something has gone wrong with the automatic creation of cross-reference. "before" and "after" are converted to links to the `before` and `after` elements which have nothing to do with it.

Solved.
 
> [...] Some hint that you can find the information there would be good.

Done.

bertfrees commented:
@PaulRambags I improved the schema a bit more and updated to OBFL example to test the schema: https://github.com/PaulRambags/obfl/compare/issues/25...snaekobbi:issues/25.

PaulRambags commented:
@bertfrees I have incorporated your changes in my branch. I noticed that the sequences in the `pre-content` and `post-content` elements cannot have value `volume` in `break-after` and `break-before` anymore, which is a good improvement. I agree with these changes.

PaulRambags commented:
@kalaspuffar  if you like you may merge this PR. Thanks!
